# Proyek Analisis Data: Nama dataset
- Nama: Luthfi Aji Wicaksono
- Email: luthfi.ajiw@gmail.com
- Id Dicoding: lihat dimana ya?

## Menentukan Pertanyaan Bisnis

- Pada musim apa sepeda paling banyak disewa sepanjang 2011-2012?
- Apakah suhu dan cuaca menjadi pertimbangan pengguna dalam menyewa sepeda?
- Tentukan preferensi waktu rental user casual dan user terdaftar
- Apakah sepanjang tahun 2011-2012 penyewaan sepeda ini mengalami kenaikan atau penurunan?

## Menyiapkan semua library yang dibutuhkan

In [1]:
import pandas as pd

## Data Wrangling

### Gathering Data

Data yang akan dipakai adalah data harian karena dianggap sudah cukup menjawab 2 pertanyaan di atas. Di bawah ini data harian rental sepeda dimuat ke dalam data frame.

In [2]:
daily_df = pd.read_csv("./Bike-sharing-dataset/day.csv")
daily_df.head()

instant      dteday  season  yr  mnth  holiday  weekday  workingday  \
0        1  2011-01-01       1   0     1        0        6           0   
1        2  2011-01-02       1   0     1        0        0           0   
2        3  2011-01-03       1   0     1        0        1           1   
3        4  2011-01-04       1   0     1        0        2           1   
4        5  2011-01-05       1   0     1        0        3           1   

   weathersit      temp     atemp       hum  windspeed  casual  registered  \
0           2  0.344167  0.363625  0.805833   0.160446     331         654   
1           2  0.363478  0.353739  0.696087   0.248539     131         670   
2           1  0.196364  0.189405  0.437273   0.248309     120        1229   
3           1  0.200000  0.212122  0.590435   0.160296     108        1454   
4           1  0.226957  0.229270  0.436957   0.186900      82        1518   

    cnt  
0   985  
1   801  
2  1349  
3  1562  
4  1600

### Assessing Data

Setelah data dimuat selanjutnya kita periksa kualitas dan permasalahan pada data yang ada. Dimulai dari memeriksa kelengkapan data dan tipe data tiap variabelnya.

In [3]:
daily_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB


Tidak ada *missing values* di semua kolom, tetapi variabel `dteday` masih berbentuk objek, nantinya data `dteday` akan dikonversi ke dalam bentuk `datetime`. Selanjutnya kita periksa apakah ada duplikasi data.

In [4]:
print("Jumlah Duplikasi: ", daily_df.duplicated().sum())

Jumlah Duplikasi:  0


Data kita sudah aman dari duplikasi, selanjutnya kita periksa apakah ada *inacccurate values* seperti *outlier* atau tidak.

In [5]:
daily_df.describe()

instant      season          yr        mnth     holiday     weekday  \
count  731.000000  731.000000  731.000000  731.000000  731.000000  731.000000   
mean   366.000000    2.496580    0.500684    6.519836    0.028728    2.997264   
std    211.165812    1.110807    0.500342    3.451913    0.167155    2.004787   
min      1.000000    1.000000    0.000000    1.000000    0.000000    0.000000   
25%    183.500000    2.000000    0.000000    4.000000    0.000000    1.000000   
50%    366.000000    3.000000    1.000000    7.000000    0.000000    3.000000   
75%    548.500000    3.000000    1.000000   10.000000    0.000000    5.000000   
max    731.000000    4.000000    1.000000   12.000000    1.000000    6.000000   

       workingday  weathersit        temp       atemp         hum   windspeed  \
count  731.000000  731.000000  731.000000  731.000000  731.000000  731.000000   
mean     0.683995    1.395349    0.495385    0.474354    0.627894    0.190486   
std      0.465233    0.544894    0.183051    0.162961    0.142429    0.077498   
min      0.000000    1.000000    0.059130    0.079070    0.000000    0.022392   
25%      0.000000    1.000000    0.337083    0.337842    0.520000    0.134950   
50%      1.000000    1.000000    0.498333    0.486733    0.626667    0.180975   
75%      1.000000    2.000000    0.655417    0.608602    0.730209    0.233214   
max      1.000000    3.000000    0.861667    0.840896    0.972500    0.507463   

            casual   registered          cnt  
count   731.000000   731.000000   731.000000  
mean    848.176471  3656.172367  4504.348837  
std     686.622488  1560.256377  1937.211452  
min       2.000000    20.000000    22.000000  
25%     315.500000  2497.000000  3152.000000  
50%     713.000000  3662.000000  4548.000000  
75%    1096.000000  4776.500000  5956.000000  
max    3410.000000  6946.000000  8714.000000

*Outlier* tampaknya tidak terlihat ada di data ini. Sepertinya beberapa nama kolom perlu diubah supaya lebih mudah untuk dibaca, serta beberapa nilai seperti; temp, atemp, dan lainnya yang perlu dikonversi agar lebih mudah dikenali.

### Cleaning Data

Pertama beberapa nama kolom perlu diganti seperti: `dteday`,`yr`, `mnth` dan `cnt`.

In [6]:
daily_df.rename(columns={
  'dteday': 'date',
  'yr': 'year',
  'mnth': 'month',
  'cnt': 'count'
}, inplace=True)
daily_df.head()

instant        date  season  year  month  holiday  weekday  workingday  \
0        1  2011-01-01       1     0      1        0        6           0   
1        2  2011-01-02       1     0      1        0        0           0   
2        3  2011-01-03       1     0      1        0        1           1   
3        4  2011-01-04       1     0      1        0        2           1   
4        5  2011-01-05       1     0      1        0        3           1   

   weathersit      temp     atemp       hum  windspeed  casual  registered  \
0           2  0.344167  0.363625  0.805833   0.160446     331         654   
1           2  0.363478  0.353739  0.696087   0.248539     131         670   
2           1  0.196364  0.189405  0.437273   0.248309     120        1229   
3           1  0.200000  0.212122  0.590435   0.160296     108        1454   
4           1  0.226957  0.229270  0.436957   0.186900      82        1518   

   count  
0    985  
1    801  
2   1349  
3   1562  
4   1600

Selanjutnya kita konversi ulang data seperti: `temp`, `atemp`, `hum` dan `windspeed` dan mengonversi tipe data date dari object menjadi `datetime`.

In [7]:
# Konversi temp ke dalam bentuk Celcius
daily_df["temp"] = daily_df["temp"].apply(lambda x: x * 41)
# Konversi atemp ke dalam bentuk Celcius
daily_df["atemp"] = daily_df["atemp"].apply(lambda x: x * 50)

daily_df["hum"] = daily_df["hum"].apply(lambda x: x * 100)
daily_df["windspeed"] = daily_df["windspeed"].apply(lambda x: x * 67)

daily_df["date"] = pd.to_datetime(daily_df["date"])

daily_df.head()

instant       date  season  year  month  holiday  weekday  workingday  \
0        1 2011-01-01       1     0      1        0        6           0   
1        2 2011-01-02       1     0      1        0        0           0   
2        3 2011-01-03       1     0      1        0        1           1   
3        4 2011-01-04       1     0      1        0        2           1   
4        5 2011-01-05       1     0      1        0        3           1   

   weathersit       temp     atemp      hum  windspeed  casual  registered  \
0           2  14.110847  18.18125  80.5833  10.749882     331         654   
1           2  14.902598  17.68695  69.6087  16.652113     131         670   
2           1   8.050924   9.47025  43.7273  16.636703     120        1229   
3           1   8.200000  10.60610  59.0435  10.739832     108        1454   
4           1   9.305237  11.46350  43.6957  12.522300      82        1518   

   count  
0    985  
1    801  
2   1349  
3   1562  
4   1600

## Exploratory Data Analysis (EDA)

**Eksplorasi Data Musim**

Pada eksplorasi ini kita mencari tahu tentang jumlah keseluruhan pengguna yang menyewa sepeda pada tiap musimnya.

In [8]:
daily_df.groupby('season').agg({
  'count': 'sum'
})

count
season         
1        471348
2        918589
3       1061129
4        841613

Pada pivot table di atas didapati bahwa pada musim:
- **Spring** terdapat **471,348** yang disewa
- **Summer** terdapat **918,589** yang disewa
- **Fall** terdapat **1,061,129** yang disewa
- **Winter** terdapat **841,613** yang disewa

**Eksplorasi Data Cuaca**

Selanjutnya mari kita temukan korelasi untuk mencari hubungan antara variabel `season`, `weathersit`, `temp` dan `count`.

In [17]:
correlation = daily_df[['season', 'weathersit', 'temp', 'count']].corr()
print(correlation)

              season  weathersit      temp     count
season      1.000000    0.019211  0.334315  0.406100
weathersit  0.019211    1.000000 -0.120602 -0.297391
temp        0.334315   -0.120602  1.000000  0.627494
count       0.406100   -0.297391  0.627494  1.000000


Korelasi di atas menunjukan bahwa **tidak ada** hubungan antara banyaknya jumlah penyewa dengan cuaca, tetapi jumlah penyewa **memiliki hubungan** moderat dengan suhu udara dan musim.

**Eksplorasi Preferensi Waktu Sewa**

Mari kita cari tahu preferensi waktu sewa pengguna pada `workingday` dan `holiday`.

In [16]:
daily_df.groupby(by='workingday').agg({
  'count': 'sum'
})

count
workingday         
0           1000269
1           2292410

Angka **0** pada `workingday` berarti bukan hari libur atau weekend dan angka **1** berarti kebalikannya. Dari situ diketahui bahwa pengguna menyewa sepeda lebih banyak pada hari kerja sebesar **2,292,410**

In [13]:
daily_df.groupby(by='holiday').agg({
  'count': 'sum'
})

count
holiday         
0        3214244
1          78435

Angka 0 pada `holiday` berarti bukan hari libur dan angka **1** berarti kebalikannya. Dari situ diketahui bahwa pengguna menyewa sepeda jauh lebih banyak pada hari biasa, bukan hari libur.

**Eksplorasi Jumlah Sewa**

Disini kita akan mencari tahu jumlah sewa pertahunnya, apakah ada peningkatan pada tahun 2012 atau tidak.

In [20]:
# Kita buat kolom baru dulu yang hanya berisi tahun
daily_df["year"] = daily_df["date"].dt.year

# Selanjutnya dibuatkan pivot table
daily_df.groupby(by="year").agg({
  'count': 'sum'
})

count
year         
2011  1243103
2012  2049576

Ternyata memang ada peningkatan jumlah sewa dalam kurun waktu satu tahun.

## Visualization & Explanatory Analysis

### Pertanyaan 1: 

### Pertanyaan 2: 

## Conclusion

- Conclution pertanyaan 1
- conclution pertanyaan 2